tensorflow/compiler/tf2tensorrt/trt_convert.cc

```diff
@@ -0,0 +1,389 @@
+  (*trt_parameter_map)["max_cached_engines"].set_i(params.max_cached_engines);
+  (*trt_parameter_map)["use_calibration"].set_b(params.use_calibration);
+  (*trt_parameter_map)["profile_strategy"].set_s(
+      ProfileStrategyToName(params.profile_strategy));
+  (*trt_parameter_map)["use_implicit_batch"].set_b(!params.use_dynamic_shape);
+  (*trt_parameter_map)["_allow_build_at_runtime"].set_b(
+      params.allow_build_at_runtime);
+  return Status::OK();
+}
+
+// Runs TRTOptimizer grappler pass.
+Status RunTfTrt(const MetaGraphDef& meta_graph_def,
+                const RewriterConfig& rewriter_config,
+                GraphDef* segmented_graph_def) {
+  ConfigProto config_proto;
+  RewriterConfig* opt_config =
+      config_proto.mutable_graph_options()->mutable_rewrite_options();
+
+  config_proto.mutable_graph_options()->mutable_rewrite_options()->CopyFrom(
+      rewriter_config);
+
+  VLOG(4) << "Setting up Grappler parameters\n" << config_proto.DebugString();
+
+  TF_RETURN_IF_ERROR(
+      RunGrappler(meta_graph_def, config_proto, segmented_graph_def));
+
+  return Status::OK();
+}
+
+Status ToggleProfileGenerationMode(GraphDef* graph_def, bool mode) {
+  VLOG(3) << "Setting _profile_generation_mode=" << mode;
+  std::string op{"TRTEngineOp"};
+  for (auto& node : *(graph_def->mutable_node())) {
+    if (!op.compare(node.op())) {
+      auto* attr = node.mutable_attr();
+      AttrValue profile_generation_mode;
+      profile_generation_mode.set_b(mode);
+      (*attr)["_profile_generation_mode"] = profile_generation_mode;
+    }
+  }
+  return Status::OK();
+}
+
+Status RunHelper(Session* session, const SignatureDef& signature,
+                 const std::vector<Tensor>& input_tensors, string prefix = "") {
+  std::vector<std::string> input_names =
+      RemoveTensorSuffix(signature.inputs(), prefix);
+  std::vector<std::string> output_names =
+      RemoveTensorSuffix(signature.outputs(), prefix);
+
+  std::vector<std::pair<std::string, tensorflow::Tensor>> input_pairs;
+  for (int i = 0; i < input_names.size(); i++) {
+    input_pairs.push_back({input_names.at(i), input_tensors.at(i)});
+  }
+  std::vector<tensorflow::Tensor> output_tensors;
+  for (int i = 0; i < output_names.size(); i++) {
+    output_tensors.push_back({});
+  }
+  VLOG(3) << "TF-TRT Build mode: running inference\n";
+  TF_RETURN_IF_ERROR(
+      session->Run(input_pairs, output_names, {}, &output_tensors));
+  return Status::OK();
+}
+
+// Runs the model to create the engines. In dynamic shape mode, before creating
+// the engines, we provide shapes to define optimization profiles.
+Status Build(GraphDef& segmented_graph_def,
+             const std::vector<std::vector<tensorflow::Tensor>>& inputs,
+             const SignatureDef& signature, Session* session,
+             const TfTrtConversionParams params) {
+  VLOG(2) << "Building the model";
+  bool profile_generation_mode = params.use_dynamic_shape && inputs.size() > 1;
```

Comment:
Let's rename profile_generation_mode to need_shape_profile.
why inputs.size() > 1? Do you mean >=1?
Also, we don't call Build if there is no input. So what we need is just this:
bool need_shape_profile = params.use_dynamic_shape